### ISSUE TITLE
Allow a throughput limit

### ISSUE THREAD

**Issue by kentquirk** [honeycombio/refinery#956]
**Is your feature request related to a problem? Please describe.**

- Throughput samplers aren't good enough for variable traffic. "What would be great would be a Dynamic mode that's the default and then Throughput mode kicks in only if we are above some threshold for overall volume."

**Describe the solution you'd like**

- Refinery monitors throughput (events per second) as a constant feature
- There's a configuration value that allows setting throughput either at the cluster level or at the individual refinery level
- It maintains a sample rate multiplier that is 1 if the total throughput is below the limit
- If the throughput value goes above the specified limit, then the sample rate multiplier starts increasing
- All sample rates generated by the selected samplers are multiplied by this throughput multiplier to bring sample rates back down to within the limit
- The current value of the limit is added as a `meta.` value.

**Describe alternatives you've considered**

- Creating special sampler variants that do this, which would require creating multiple new sampler types.

**Additional context**

- Partially discussed in #525 
- Also discussed in [Pollinators](https://honeycombpollinators.slack.com/archives/C01BZ3J7RMJ/p1703010013565929)

**Comment by VinozzZ** [comment]
I would like to work on this. To ensure I understand both the targeted problem and the proposed solution, I’m going to reiterate them here.

Problem:
Customers want to maintain a consistent sample rate for both bursty traffic and low-volume traffic, ensuring that each type receives equal representation in the sampled data. However, when an unexpected surge of bursty traffic occurs, customers still need to keep the overall event volume under control. This means the sample rate needs to dynamically adjust in response to unexpected traffic bursts, thereby enforcing a total throughput limit.

Proposed Solution:
We propose introducing a new configuration option, ThroughputLimit, in the DynamicSampler. If this option is set, a sample rate multiplier will be applied based on the specified limit and the current throughput. When the current throughput exceeds the limit, the sample rate will be adjusted using this multiplier to bring the throughput back within the defined limit.

Please let me know if this makes sense or if I have misunderstood any aspects.

**Comment by kentquirk** [comment]
I agree with what you said, except for where it is and what it affects -- the configuration option I'm talking about would apply to Refinery as a whole, and would multiply all calculated sample rates from any sampler configuration.

This means that each Refinery in a cluster needs to calculate its output events per second using a moving average based on the number of spans being sent and gossip that periodically. Like stress level, all the refineries in the cluster should share the same calculated value and use it together.

There probably also have to be a couple more configuration parameters with useful defaults -- like AdjustmentInterval and Weight.

**Comment by VinozzZ** [comment]
If the original sample rate is 1, the multiplier should by default not modify those sample rate

**Comment by VinozzZ** [comment]
After running some tests with the initial implementation, the feature is more complicated and requires more work than we planned. We are going to give this feature work more time and not include it in the upcoming 2.8 release.
